pramen/core/src/main/scala/za/co/absa/pramen/core/utils/JdbcSparkUtils.scala

```diff
@@ -77,24 +77,110 @@ object JdbcSparkUtils {
+    * @param schema        A schema.
+    * @param tableName     The name of the table in the database. It can contain schema name and catalog name.
+    * @param connection    A JDBC connection to the database engine.
+    * @return The schema with column descriptions added as metadata.
+    */
+  def addColumnDescriptionsFromJdbc(schema: StructType, tableName: String, connection: Connection): StructType = {
+    val fieldToDescriptionMap: Map[String, String] = try {
+      val columns = getColumnMetadata(tableName, connection)
+      val fieldsDescription = new ListBuffer[(String, String)]
+
+      while (columns.next()) {
+        val columnNameOpt = Option(columns.getString("COLUMN_NAME"))
+        val descriptionOpt = Option(columns.getString("REMARKS"))
+
+        (columnNameOpt, descriptionOpt) match {
+          case (Some(columnName), Some(description)) =>
+            fieldsDescription += columnName.toLowerCase -> description
+          case _ =>
+        }
+      }
+      columns.close()
+      fieldsDescription.toMap[String, String]
+    } catch {
+      case ex: Throwable =>
+        log.warn(s"Unable to fetch metadata for database table: $tableName", ex)
+        return schema
+    }
+
+    StructType(
+      schema.fields.map {
+        field =>
+          val fieldNameLowerCase = field.name.toLowerCase
+
+          fieldToDescriptionMap.get(fieldNameLowerCase) match {
+            case Some(description) =>
+              val metadata = new MetadataBuilder
+              metadata.withMetadata(field.metadata)
+              metadata.putString(COMMENT_METADATA_KEY, description)
+              field.copy(metadata = metadata.build())
+            case _ =>
+              field
+          }
+      }
+    )
+  }
+
+  /**
+    * Gets metadata of table columns for database engines that support it.
+    *
+    * @param fullTableName  The name of the table in the database. It can contain schema name and catalog name.
+    * @param connection     A JDBC connection to the database engine.
+    * @return The ResultSet containing metadata for all columns.
+    */
+  def getColumnMetadata(fullTableName: String, connection: Connection): ResultSet = {
+    val dbMetadata: DatabaseMetaData = connection.getMetaData
+
+    if (!dbMetadata.getColumns(null, null, fullTableName, null).next()) {
+      val parts = fullTableName.split('.')
+      if (parts.length == 3) {
+        // database, schema, and table table are all present
+        dbMetadata.getColumns(parts(0), parts(1), parts(2), null)
+      } else if (parts.length == 2) {
+        if (dbMetadata.getColumns(null, parts(0), parts(1), null).next()) {
+          dbMetadata.getColumns(null, parts(0), parts(1), null)
+          // schema and table only
+        } else {
+          // database and table only
+          dbMetadata.getColumns(parts(0), null, parts(1), null)
+        }
+      } else {
+        // table only
+        dbMetadata.getColumns(null, null, fullTableName.toUpperCase, null)
```

Comment:
For some reason when a table is created in HSQL, all database names, table names, and column names are uppercased. `dbMetadata.getColumns(null, null, fullTableName, null)` didn't work for a lowercase table name.
The exact casing was covered at line 143, and no metadata was found, so trying uppercased name to increase chances.

This made me think, I should check lowercase too in case the db engine makes columns lowercase and inputs to `dbMetadata.getColumns` are case sensitive.

Reply:
Here is an example metadata returned by dbMetadata.getColumns():
```
TABLE_CAT	TABLE_SCHEM	TABLE_NAME	COLUMN_NAME	 DATA_TYPE
PUBLIC	        PUBLIC	        COMPANY	        ID               4	        
PUBLIC     	PUBLIC	        COMPANY	        NAME             12	       
PUBLIC	        PUBLIC	        COMPANY	        DESCRIPTION      12	       
PUBLIC	        PUBLIC	        COMPANY	        EMAIL            12	       
```